### DISCUSSION TITLE
Using atoti Without Open Ports

### DISCUSSION THREAD

**Discussion by shahin-y** [atoti/atoti#759] [Q&A] [answered]
Hi everyone,

How can we run atoti in an environment without open ports for it?

We are facing this scenario when Jupyter Lab is served via a container (think of a Jupyter Hub or a Cloudera Data Science Workbench setup) and there aren't any other open ports available.

```python
import atoti as tt
session = tt.Session()

session.link()
```
Result: `Cannot access the session after trying this URL: http://<url>:<port>`

Not being able to get the (external) link is reasonable I guess. 
Unfortunately, the jupyterlab widget also does not work in the notebook.
 ```python
session.visualize()
```
This just seems to block the kernel and the widget can't connect.
On the other hand, querying the session via `localhost` works:
```python
import requests

response = requests.get(
        f"http://localhost:{session.port}/",
        headers={"Content-Type": "application/json", "Accept": "application/json"}
    )
response
```
Result: `<Response [200]>`

Does the jupyterlab extension maybe not support this setup?

Thanks for any help!

**Comment by tibdex** [comment, marked as answer]
https://github.com/jupyterhub/jupyter-server-proxy is made exactly for that. We briefly mention it in [our docs](https://docs.atoti.io/0.7.3/lib/atoti-query/atoti_query.query_session.html#atoti_query.QuerySession.link). Once enabled, both `Session.link()` and `Session.visualize()` should work.

**Reply by shahin-y** [reply]
Thanks for your prompt reply @tibdex! 

I missed that. This should work 👍
